coderd/httpmw/apikey.go

```diff
@@ -446,6 +462,207 @@ func ExtractAPIKey(rw http.ResponseWriter, r *http.Request, cfg ExtractAPIKeyCon
+	if audienceURI == "" {
+		return ""
+	}
+
+	u, err := url.Parse(audienceURI)
+	if err != nil {
+		// If parsing fails, return as-is to avoid breaking existing functionality
+		return audienceURI
+	}
+
+	// Apply RFC 3986 syntax-based normalization:
+
+	// 1. Scheme normalization - case-insensitive
+	u.Scheme = strings.ToLower(u.Scheme)
+
+	// 2. Host normalization - case-insensitive and IDN (punnycode) normalization
+	u.Host = normalizeHost(u.Host)
+
+	// 3. Remove default ports for HTTP/HTTPS
+	if (u.Scheme == "http" && strings.HasSuffix(u.Host, ":80")) ||
+		(u.Scheme == "https" && strings.HasSuffix(u.Host, ":443")) {
+		// Extract host without default port
+		if idx := strings.LastIndex(u.Host, ":"); idx > 0 {
+			u.Host = u.Host[:idx]
+		}
+	}
+
+	// 4. Path normalization including dot-segment removal (RFC 3986 Section 6.2.2.3)
+	u.Path = normalizePathSegments(u.Path)
+
+	// 5. Remove fragment (OAuth2 audience should not have fragments anyway)
+	u.Fragment = ""
+
+	// 6. Keep query parameters as-is (rarely used in audience URIs but preserved for compatibility)
+
+	return u.String()
+}
+
+// normalizeHost performs host normalization including case-insensitive conversion
+// and IDN (Internationalized Domain Name) punnycode normalization.
+func normalizeHost(host string) string {
+	if host == "" {
+		return host
+	}
+
+	// Handle IPv6 addresses - they are enclosed in brackets
+	if strings.HasPrefix(host, "[") && strings.HasSuffix(host, "]") {
+		// IPv6 addresses should be normalized to lowercase
+		return strings.ToLower(host)
+	}
+
+	// Extract port if present
+	var port string
+	if idx := strings.LastIndex(host, ":"); idx > 0 {
+		// Check if this is actually a port (not part of IPv6)
+		if !strings.Contains(host[idx+1:], ":") {
+			port = host[idx:]
+			host = host[:idx]
+		}
+	}
+
+	// Convert to lowercase for case-insensitive comparison
+	host = strings.ToLower(host)
+
+	// Apply IDN normalization - convert Unicode domain names to ASCII (punnycode)
+	if normalizedHost, err := idna.ToASCII(host); err == nil {
+		host = normalizedHost
+	}
+	// If IDN conversion fails, continue with lowercase version
+
+	return host + port
+}
+
+// normalizePathSegments implements RFC 3986 Section 6.2.2.3 dot-segment removal
+// and other path normalizations for consistent OAuth2 audience matching.
+func normalizePathSegments(path string) string {
+	if path == "" {
+		// If no path is specified, use "/" for consistency with RFC 8707 examples
+		return "/"
+	}
+
+	// First, resolve dot segments as per RFC 3986
+	path = resolveDotSegments(path)
+
+	// Remove trailing slash from paths longer than "/" to normalize
+	// This ensures "/api/" and "/api" are treated as equivalent
+	if len(path) > 1 && strings.HasSuffix(path, "/") {
+		path = strings.TrimSuffix(path, "/")
+	}
+
+	return path
+}
+
+// resolveDotSegments implements RFC 3986 Section 5.2.4 algorithm
+// for removing dot segments from a path.
+func resolveDotSegments(path string) string {
```

Comment:
`url.URL.ResolveReference()` should do this for us? 

https://pkg.go.dev/net/url#URL.ResolveReference

Reply:
done